substrate/frame/referenda/src/lib.rs

```diff
@@ -1255,22 +1278,35 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		who: T::AccountId,
 		amount: BalanceOf<T, I>,
 	) -> Result<Deposit<T::AccountId, BalanceOf<T, I>>, DispatchError> {
-		T::Currency::reserve(&who, amount)?;
+		T::Currency::hold(&HoldReason::Referendum.into(), &who, amount)?;
 		Ok(Deposit { who, amount })
 	}
 
 	/// Return a deposit, if `Some`.
-	fn refund_deposit(deposit: Option<Deposit<T::AccountId, BalanceOf<T, I>>>) {
+	fn refund_deposit(
+		deposit: Option<Deposit<T::AccountId, BalanceOf<T, I>>>,
+	) -> Result<BalanceOf<T, I>, DispatchError> {
 		if let Some(Deposit { who, amount }) = deposit {
-			T::Currency::unreserve(&who, amount);
+			T::Currency::release(&HoldReason::Referendum.into(), &who, amount, Precision::Exact)
+		} else {
+			Ok(Zero::zero())
 		}
 	}
 
 	/// Slash a deposit, if `Some`.
 	fn slash_deposit(deposit: Option<Deposit<T::AccountId, BalanceOf<T, I>>>) {
 		if let Some(Deposit { who, amount }) = deposit {
-			T::Slash::on_unbalanced(T::Currency::slash_reserved(&who, amount).0);
-			Self::deposit_event(Event::<T, I>::DepositSlashed { who, amount });
+			let (imbalance, non_slashed) = <T::Currency as FnBalanced<T::AccountId>>::slash(
+				&HoldReason::Referendum.into(),
+				&who,
+				amount,
+			);
+			T::OnSlash::on_unbalanced(imbalance);
+
+			Self::deposit_event(Event::<T, I>::DepositSlashed {
+				who,
+				amount: amount.saturating_sub(non_slashed),
```

Comment:
I think its possible to first emit the event and use `imbalance.peek()`, that makes it a bit more understandable IMHO.